### DISCUSSION TITLE
Cannot connect to an NTP server

### DISCUSSION THREAD

**Discussion by wsw70** [authelia/authelia#4393] [Q&A] [answered]
I configured Authelia to connect to my NTP server (`chrony`):

```yaml
ntp:
  ## NTP server address.
  address: "192.168.10.1:123"
```

The NTP server is live and not firewalled:

```
root@srv # nmap -sU -p 123 192.168.10.1

PORT    STATE SERVICE
123/udp open  ntp
```

It is also reachable from the Authelia docker container:

```
root@srv ~# docker exec -it authelia-authelia-1 ping 192.168.10.1
PING 192.168.10.1 (192.168.10.1): 56 data bytes
64 bytes from 192.168.10.1: seq=0 ttl=63 time=0.178 ms
```

Despite all of this, Authelia cannot use it:

```
authelia-authelia-1  | time="2022-11-16T21:04:22+01:00" level=info msg="Authelia v4.37.2 is starting"
authelia-authelia-1  | time="2022-11-16T21:04:22+01:00" level=info msg="Log severity set to info"
authelia-authelia-1  | time="2022-11-16T21:04:22+01:00" level=info msg="Storage schema is being checked for updates"
authelia-authelia-1  | time="2022-11-16T21:04:22+01:00" level=info msg="Storage schema is already up to date"
authelia-authelia-1  | time="2022-11-16T21:04:27+01:00" level=warning msg="Could not read from the NTP server socket to validate the system time is properly synchronized: read udp 172.18.0.26:34486->192.168.10.1:123: i/o timeout"
authelia-authelia-1  | time="2022-11-16T21:04:27+01:00" level=info msg="Initializing server for non-TLS connections on '[::]:9091' path '/'"
```

What could be the reason for that?

**Comment by james-d-elliott** [comment, marked as answer]
The i/o timeout is a low level error indicating a port is not responding in a timely manner. Everything past `synchronized:` is from the go stdlib.

You can try the following:

```
docker exec -it authelia /bin/ash
```

```
apk add --no-cache ntpsec
ntpq -p 192.168.10.1
```

**Reply by wsw70** [reply]
> The i/o timeout is a low level error indicating a port is not responding in a timely manner. Everything past `synchronized:` is from the go stdlib.

Thank you for your answer. I am not an expert in NTP (I just use it as an end user on my systems) but I tried to do some testing. 

- All the NTP queries below happen from within the container. 
- The docker host is `192.168.10.2`.  This is the place I run the `tcpdump` commands below.
- The NTP server (running `chrony`) is `192.168.10.1`.

First of all,  `192.168.10.2` itself is successfully synchronized with  `192.168.10.1`. 

I ran the following Python script from within the container. I wanted to make sure that the  `192.168.10.1` server is not only reachable, but also correctly responds

```
import ntplib
from time import ctime
c = ntplib.NTPClient()
response = c.request('192.168.10.1', version=3)
print(ctime(response.tx_time))
response = c.request('192.168.10.1', version=4)
print(ctime(response.tx_time))
```

The response is 
```
/ # python3 testntp.py
Thu Nov 17 14:50:41 2022
Thu Nov 17 14:50:41 2022
```

So the NTP traffic is OK. From the server ( `192.168.10.1`) perspective:

```
15:51:36.455828 enp3s0 In  IP 192.168.10.2.45198 > 192.168.10.1.123: NTPv3, Client, length 48
15:51:36.455891 enp3s0 Out IP 192.168.10.1.123 > 192.168.10.2.45198: NTPv3, Server, length 48
15:51:36.456344 enp3s0 In  IP 192.168.10.2.34076 > 192.168.10.1.123: NTPv4, Client, length 48
15:51:36.456366 enp3s0 Out IP 192.168.10.1.123 > 192.168.10.2.34076: NTPv4, Server, length 48
```

There is a packet `Client` sent, and a reply in `Server` (twice for two versions of NTP).

Now to your command:

```
/ # ntpq -p 192.168.10.1
***Request timed out
```

On the server:

```
15:52:59.115083 enp3s0 In  IP 192.168.10.2.34398 > 192.168.10.1.123: NTPv3, Control Message, length 12
15:53:04.118933 enp3s0 In  IP 192.168.10.2.34398 > 192.168.10.1.123: NTPv3, Control Message, length 12
```

This is one line of a request that does not seem to be correctly processed by `chrony`. Please note that the packet is now `Control Message`.

I do not really know where to go from there. It looks like the server is OK and correctly reachable from within the container, but what is being sent is not what is expected.

**Reply by james-d-elliott** [reply]
I would recommend disabling the startup check until you figure it out.

The only thing I can think of is if there is some strange networking issue which you could test for by ensuring the authelia container can retrieve something else from the server via something like http.

**Reply by wsw70** [reply]
> The only thing I can think of is if there is some strange networking issue which you could test for by ensuring the authelia container can retrieve something else from the server via something like http.

Connectivity works fine (commands started from the docker hosts and executed from within the Authelia container):

```
root@srv ~# docker exec -it authelia-authelia-1 wget -O - https://www.google.com/robots.txt
Connecting to www.google.com (172.217.18.196:443)
writing to stdout
User-agent: *
Disallow: /search
Allow: /search/about
Allow: /search/static
Allow: /search/howsearchworks
(...)
```

Same for connectivity to the host where the NTP service is (what responds is Pi-hole):

```
root@srv ~# docker exec -it authelia-authelia-1 wget -O - http://192.168.10.1
Connecting to 192.168.10.1 (192.168.10.1:80)
writing to stdout
    <!doctype html>
    <html lang='en'>
        <head>
(...)
```

**Reply by james-d-elliott** [reply]
Why is PiHole responding? Is it on the same host? If you stop the web service on the host with NTP does the http request fail?

**Reply by wsw70** [reply]
> Why is PiHole responding? Is it on the same host? 

Yes, it is the same host (a router that gathers the key "logistical" functionality of my network: DNS, DHCP, NTP, ...). Pi-hole is one of the services and the only one that needs port 80 (and 53, and 6x for DHCP)

> If you stop the web service on the host with NTP does the http request fail?

Yes (I am not sure to understand the question).

I will have a closer look at Authelia's code for NTP and test from there.

**Reply by james-d-elliott** [reply]
What NTP sofrware/dist are you running?

**Reply by wsw70** [reply]
> What NTP sofrware/dist are you running?

Chrony (v4.0) on a Debian

**Reply by james-d-elliott** [reply]
Can't replicate this with authelia on 3.x/4.2/4.3 on various OS's (Debian, Ubuntu, AlmaLinux, Alpine), worked fine. As soon as I install it and configure chronyd to allow clients from the relevant networks it just works with authelia.

**Reply by james-d-elliott** [reply]
Try with the default settings (completely commented causes authelia to use cloudflare), if they work without an error/warning it's almost certainly a chronyd issue. If it doesn't it's likely to be a networking issue.

**Reply by xsolinsx** [reply]
for the records:
I had the same problems when upgrading from 4.36.9 to 4.37.5 (i/o timeout when connecting to the smtp server or the ntp server)
my issue was that the /etc/resolv.conf inside the **pihole container** was pointing to 127.0.0.11 (weird I know, probably due to an update to the pihole container or IDK)
I created a bind-mount to a local file for the pihole container in which there is only this line
`nameserver 127.0.0.1`
now everything is working fine

**Reply by wsw70** [reply]
@xsolinsx your comment made me revisit the problem and, surprise, it is gone (authelia is now at 4.37.5). the only major change I put in place is that I enabled IPv6 on my LAN. This should not have made a difference as the call to the NTP server is on IPv4 but who knows?

As for `127.0.0.11` - this is (or used to be) the standard IP on which the DNS server was bound in some configurations. As much as I love docker (and have all my services running on docker), the DNS part is really weird and unpredictable, and documentation changes from place to place. It works after some voodoo, the right number of restarts and the proper alignment of planets - and then I keep my fingers crossed so that it does not go down the drain at the next reconfiguration or major upgrade.
